backend/typescript/rest/checkInRoutes.ts

```diff
@@ -0,0 +1,103 @@
+import { Router } from "express";
+
+import ICheckInService from "../services/interfaces/checkInService";
+import CheckInService from "../services/implementations/checkInService";
+import { sendResponseByMimeType } from "../utilities/responseUtil";
+import getErrorMessage from "../utilities/errorMessageUtil";
+
+const checkInRouter: Router = Router();
+
+const checkInService: ICheckInService = new CheckInService();
+
+checkInRouter.get("/:id?", async (req, res) => {
+  const { id } = req.params;
+  const { volunteerId } = req.query;
+  const contentType = req.headers["content-type"];
+
+  if (id && volunteerId) {
+    await sendResponseByMimeType(res, 400, contentType, [
+      {
+        error: "Cannot get by both id and volunteerId",
+      },
+    ]);
+    return;
+  }
+  if (!id && !volunteerId) {
+    try {
+      const checkIns = await checkInService.getAllCheckIns();
+      res.status(200).json(checkIns);
+    } catch (error: unknown) {
+      res.status(500).json({ error: getErrorMessage(error) });
+    }
+  } else if (id) {
+    try {
+      const checkIns = await checkInService.getCheckInsById(id);
+      res.status(200).json(checkIns);
+    } catch (error: unknown) {
+      res.status(500).json({ error: getErrorMessage(error) });
+    }
+  } else if (volunteerId) {
+    try {
+      const checkIns = await checkInService.getCheckInsByVolunteerId(
+        volunteerId as string,
+      );
+      res.status(200).json(checkIns);
+    } catch (error: unknown) {
+      res.status(500).json({ error: getErrorMessage(error) });
+    }
+  }
+});
+
+/* Delete checkins by id (e.g. /checkin/63)
+  or deletes by start and end date range
+  (e.g. /checkin?startDate=2022-01-31T05:00:00.000Z?endDate=2022-02-31T05:00:00.000Z)
+*/
+checkInRouter.delete("/:id?", async (req, res) => {
+  const { id } = req.params;
+  const { startDate, endDate } = req.query;
+  const contentType = req.headers["content-type"];
+
+  if (id && (startDate || endDate)) {
+    await sendResponseByMimeType(res, 400, contentType, [
+      {
+        error: "Cannot delete by both id and startDate/endDate",
+      },
+    ]);
+    return;
+  }
+
+  if (id) {
+    try {
+      await checkInService.deleteCheckInById(id);
+      res.status(204).send();
+    } catch (error: unknown) {
+      res.status(500).json({ error: getErrorMessage(error) });
+    }
+  }
+
+  if (!(startDate && endDate)) {
+    await sendResponseByMimeType(res, 400, contentType, [
+      {
+        error: "Must supply both startDate and endDate query parameters",
+      },
+    ]);
+    return;
+  }
+
+  if (startDate && endDate) {
+    if (startDate > endDate) {
+      throw new Error(`start date must be before end date.`);
+    }
+    try {
```

Comment:
should also check if they are valid date strings.

I dont think this is working right now as well because when I make the following postman call, the call just keeps loading.
<img width="1004" alt="image" src="https://user-images.githubusercontent.com/37384882/156899294-176ce3dc-9c33-467f-aa20-ce019d8b40b9.png">

I think it is because we are comparing strings, not date, so i think you should convert them to date first and then make those checks of whether they are valid dates and if start date is before end date.

Reply:
just fixed it hopefully; now we're comparing dates instead of strings by converting them first